### ISSUE TITLE
Add `insertBlankLineAtStartOfScope` rule

### ISSUE THREAD

**Issue by ruslanskorb** [nicklockwood/SwiftFormat#261]
There is a rule for [removing blank lines at the start of scope](https://github.com/nicklockwood/SwiftFormat/blob/master/Sources/Rules.swift#L583-L611).

Would you like to have a partially opposite rule?

[Some teams](https://github.com/eure/swift-style-guide#open-braces--for-type-declarations-functions-and-closures-should-be-followed-by-one-empty-line-single-statement-closures-can-be-written-in-one-line) find this useful.

- [implementation](https://github.com/rsk-lab/SwiftFormat/blob/add-insert-blank-line-rule/Sources/Rules.swift#L583-L611)
- [tests](https://github.com/rsk-lab/SwiftFormat/blob/add-insert-blank-line-rule/Tests/RulesTests.swift#L1219-L1234)

**Comment by ruslanskorb** [comment]
Alternatively, this rule can be part of the `blankLinesAtStartOfScope` rule. In this case it may be configured using `--insert` and `--remove` options. And there will be no need to [change the name of the `blankLinesAtStartOfScope` rule](#260).

**Comment by rmvz3** [comment]
I'd like to be able to insert a blank line at the start of scopes too

**Comment by dfreniche** [comment]
We use that style in our projects. Would be super-nice to have it

**Comment by dfreniche** [comment]
@nicklockwood 1st: thanks for this awesome tool and all the effort you've put in it. Much appreciated.

This has been requested several times (for instance, [here](https://github.com/nicklockwood/SwiftFormat/issues/254)). So it's safe to assume some teams will need to _insert_ a whitespace instead of removing it.

Looks like @ruslanskorb has already written the code, even tested it. Will you be OK with a PR?

Thanks!

**Comment by serges147** [comment]
@nicklockwood Just an idea: what about having dedicated `blankLinesAtStartOfXxxScope` rules, where "Xxx" could be Type|Func|Enum|Collection|Statement ?

- Type: for protocols, classes, structs
- Func: for functions and closures
- Enum: for enums
- Collection: for `[]` and `[:]`
- Statement: for things like `defer`, `for`, `while` and so on...

The existing `blankLinesAtStartOfScope` will be still in charge, but there will be possibility to override specific type of scope.

With addition of corresponding options (with same name as option?), like `--blankLinesAtStartOfEnumScope 1` for example, it would be possible both either remove or append required number of blank lines specifically for enums. Having default value = 0 it should keep current behavior.

And the same idea for the `blankLinesAtEndOfScope`.

**Comment by serges147** [comment]
I just wanna share with you guys my experience with `swiftformat` tool when I just ran it on our 1K+ swift files project. The tool is awesome! A lot of issues and inconsistencies in the code reformatted quick and nice. So, thank you very much!!! The only problem is these pair of start/end rules. It would be bummer to disable them.

**Comment by nicklockwood** [comment]
> blankLinesAtStartOfXxxScope rules, where "Xxx" could be Type|Func|Enum|Collection|Statement ?

Yeah, this could work. There'd be a lot of duplication between the logic, but maybe they could share a common implementation.

> With addition of corresponding options (with same name as rule?), like --blankLinesAtStartOfEnumScope 1 for example

This part is slightly trickier. Arguments currently need to be short and lowercase to match the existing style. I could break from the existing style, but that would be a significant change that I'd prefer not to make unless absolutely necessary.

Rules can be disabled by default though, so new rules could default to off.

They would need some kind of config though to indicate if the blank line should be inserted or removed, and having a separate rule for insert/remove isn't viable because rules can't be mutually exclusive, since there's no way to prevent the user enabling both of them.

**Comment by serges147** [comment]
Sorry, I didn't get the 
> ... if the blank line should be inserted or removed, ...

My understanding is that, having an option with exact number blank lines required, the rule can do both. Let's say option configured to `X` number.
- If scope has N blank lines (less than X) then the rule will **append** X-N lines.
- If scope has N blank lines (more than X) then the rule will **remove** N-X lines.
- Nothing changed if source has exactly X blank lines

Please correct me if I'm missing something.

> This part is slightly trickier. Arguments currently need to be short and lowercase to match the existing style.

Yup, I see what you mean. OK, another crazy idea: value of an option could be more complex then just number - json for example. In such case you can have simple and generic `--scope` new option, and all the complexity and varieties of different types and their values will be in the json. Like:
```
{
    "enum": {
        "blankLinesAtStart":1,
        "blankLinesAtEnd":0
    },
    "type": { ...
```

**Comment by nicklockwood** [comment]
> Yup, I see what you mean. OK, another crazy idea: value of an option could be more complex then just number

This isn't *that* crazy. I already support comma-delimited option arguments (arrays, basically). Full JSON would be too much, but maybe adding another delimiter for key-value pairs would be doable.

**Comment by serges147** [comment]
Key-value pairs will do, but with swift's `Codable` JSON is just piece of cake. Plus, it will allow keep things hierarchical instead of flat.

**Comment by nicklockwood** [comment]
Right, but encoding json into a command-line argument isn't really viable, since you'd have to escape double quotes.
